Polar path documentation (#319)
KILTprotocol/docs: docs/develop/09_polarpath/01_overview.md

---
id: what-is-polarpath
title: Overview
---

> Polar path consists of an `asset-swap` pallet for KILT chain runtimes and extrinisics that let parachain-based chain developers to make their native token accessible on the Ethereum network using [the Snowbridge bridge](https://docs.snowbridge.network) between Polkadot and Ethereum.

Reviewer: ```suggestion
Polar path consists of an `asset-switch` pallet for KILT chain runtimes and extrinisics that let parachain-based chain developers to make their native token accessible on the Ethereum network using [the Snowbridge bridge](https://docs.snowbridge.network) between Polkadot and Ethereum.
```

Reviewer: I feel here you're mixing the KILT deployment of polar path with the generic polar path solution, no?

Author: I don't know… If you mean what do we want to document, then is that generic for other runtimes to use?

Reviewer: Well, that's kinda my question to you. I think the KILT docs should describe what our deployment does. The pallet docs describe what the pallet does in a generic way.

Author: Hmm, OK, I may do a little bit of both. Explain broadly, and then use KILT as the example?

Polar path provides parachains with a solution for creating an ERC-20 wrapper around their native tokens, enabling a trustless conversion of any existing native parachain token (PARA) into a wrapped parachain token (wPARA) on Ethereum while maintaining the trustlessness guarantees Snowbridge provides.

> The pallet enables a configured conversion rate of parachain tokens on the origin chain to an ERC-20 wrapper token on Polkadot's AssetHub.

Reviewer: False, this is not happening anymore. It's always 1:1.

Author: I am not sure what you're referring to? Most of my source material was old, so if you could point me to any more up-to-date content that explains what you mean, that would be appreciated.

Reviewer: It's in the pallet docs. But anyway, the pallet only allows 1:1 exchanges. We scrapped the configurable ratio because of rounding issues and also because we are not sure it's something people need. If they do, we'll rethink how to do that.

## Setup

:::tip Asset Hub
The Polkadot Relay Chain doesn't natively support assets apart from DOT. Instead, parachains support this functionality, and for Polkadot, this parachain is the [Asset Hub](https://wiki.polkadot.network/docs/build-integrate-assets).
:::

Making a Parachain Token to be usable on Ethereum requires the following steps:

> 1. Deploy an ERC-20 contract token to Ethereum with a fixed supply based on the parachain needs.

Reviewer: Maybe add that in most cases the total supply should match the parachain max supply, if possible to determine (e.g., no infinite inflation).

2. [Disable the token issuer rights](https://ethereum.org/en/guides/how-to-revoke-token-access/) on the ERC-20 contract.

> 3. [Register the ERC-20 token on Asset Hub](https://wiki.polkadot.network/docs/learn-assets#creation-and-management).

Reviewer: Wrong link. Registration on AH happens via the bridge: https://docs.snowbridge.network/applications/token-transfers#token-registration.

Author: @ntn-x2 ?

```suggestion
3. [Register the ERC-20 token on Asset Hub](https://docs.snowbridge.network/applications/token-transfers#token-registration).
```

Reviewer: Looks good!

> 4. The entity holding ERC-20 funds performs a one-time transfer of all tokens from their Ethereum account to the parachain's sovereign account on Asset Hub.

Reviewer: Minus the amount of tokens that he previously locked on KILT. Those will stay under an account of his choice on Etheruem.

Author: I assume you're referring to point 4?

Questions…

1. Where in the flow does this locking happen? I don't seem to have seen that in any source material.
2. Are these tokens locked on KILT even when other chains use the pallets?

```suggestion
4. The entity holding ERC-20 funds performs a one-time transfer of all tokens from their Ethereum account minus the amount of tokens previously locked with KILT to the parachain's sovereign account on Asset Hub.
```

Reviewer: 1. The requirement in the pallet is that before a switch pair can be created, the account for that pair must have sufficient balance. This means that someone has to send enough KILTs to this account. It's not specified anywhere other than the code, which is why I think it's extremely important to describe the process.
2. No, each chain would hold their native token locked in a poo-specific account. KILT will only lock KILTs.

5. Parachain governance recognizes the wrapper around the ERC-20 token by pairing the ERC-20 token to the parachain token in the swap pallet and enable the token issuer rights.

Holders of the token on the parachain can can now switch tokens between it and Ethereum.

## Use cases

After setup, the following flows are possible:

:::tip XCM
Polar path uses Polkadot's [Cross-Consensus Message Format (XCM)](https://wiki.polkadot.network/docs/learn-xcm), a messaging format and language used to communicate between consensus systems.
:::

### Sending tokens from the parachain to Ethereum

1. Transmit custom parachain tokens, "PARA", from the parachain to wrapped tokens "wPARAs" on Asset Hub via XCM
2. Send wPARAs from Asset Hub to the ERC-20 contract on Ethereum via Snowbridge

### Sending tokens from Ethereum to the parachain

1. Send wPARAs from the ERC-20 contract on Ethereum to Asset Hub via Snowbridge
2. Transmit wPARAs on Asset Hub to PARAs on the parachain via XCM

## Extrinsics

> Before performing a switch, you first need to create a switch pair between the parachain token and the ERC-20 token using the `setSwitchPair` extrinisc. Then you can use any of the other extriniscs with the switch pair, including performing the switch itself using the `switch` extrinisc.

Reviewer: It's not really you, right? It's a configurable origin, and in our case it will be a public referendum.

Author: I'm a bit confused… Someone has to start the process yes? Someone needs to create the switch pair?

Or are there two parts to this?

Someone creates the switch pair and specifies who else can use the other extrinisics.

And what's the referendum? I've not seen anything mentioned in any of the source material about how any of the pallets refer to a referendum. Do you mean that the referendum triggers the other extrinsic or something else?

Reviewer: A new switch pair can be created by the origin configured. This is generic. In our case, it will be the `Root` origin, which is only possible to use after a referendum. So yes, someone proposes a referendum that would dispatch a specific function (in this case the switch pair creation and enactment), and if it goes through (after ~7 days), the call is dispatched, and the switch pair is created.

:::tip Remote asset
A remote asset is the identifier of the asset considered to be the other side of the switch pair. For an ERC20 token, this is a [MultiLocation](https://wiki.polkadot.network/docs/learn/xcm/fundamentals/multilocation-summary) (in XCM terms) pointing to an address on one of the many EVM-based deployments.
:::

> ### `setSwitchPair`

Reviewer: We can easily link (or even better, embed) the README from the chain side. Temporarily using a link to the PR, and eventually using the tag name of the version we will release.

Author: Do you mean instead of explaining the extriniscs? When I wrote this, nothing was available, so I wasn't sure what it would be. I'm going to wait and see what that looks like first. A higher-level overview might still be useful before diving right in, or at least I can then link to the specific parts in the file that's pulled in.

Reviewer: Alright, but now we have more docs on the chain side, although I am not sure it is at the right level of details for this docs. They might go too much in detail.
